File: cracking/Ruby/1.5.cracking.rb
=begin
write a method to replace all spaces in a string with '%20'
=end

def replace(str)
	count = 0
	str.each_char do |char|
		count += 1 if char == ' '
	end
	index = str.length - 1
	str += " " * 2 * count
	runner = str.length - 1
	index.downto(0).each do |i|
		if str[i] == ' '
			str[runner] = '0'
			str[runner - 1] = '2'
			str[runner - 2] = '%'
			runner -= 3
		else
			str[runner] = str[i]
			runner -= 1
		end
	end
	str
end

str = 'haha hoho hehe'
puts replace(str)

File: cracking/C/1.5.cracking.c
/*
Replace all the spaces in a string with %20
*/

#include <stdio.h>
#include <ctype.h>

void replace(char *s, int len, char* t) {
  int i, numOfSpaces;
  printf("i is: %d\n", i);
  printf("numOfSpaces: %d\n", numOfSpaces);
  for (i=0; i<len; i++) {
    if (isspace(s[i])) {
      numOfSpaces++;
    }
  }
  printf("numOfSpaces: %d\n", numOfSpaces);
  int runner = 0;
  while (*s) {
    if (isspace(*s)) {
      t[runner++] = '%';
      t[runner++] = '2';
      t[runner++] = '0';
    } else {
      t[runner++] = *s;
    }
    s++;
  }
  t[runner] = '\0';
}

int main() {
  char s[30] = "I love you";
  char t[30];
  replace(s, 10, t);
  printf("modified string is: %s\n", t);
  return 0;
}
